Use production-grade webserver

v0d1ch opened the pull request:
fix #806 

## Why

We want to use a production-grade webserver in hydra-node

## What

Use `wai-websockets`  and `warp` package to achieve this

---

<!-- Tick off or strike-through / remove if not applicable -->
* [x] CHANGELOG updated
* [ ] Documentation updated
* [x] Added and/or updated haddocks
* [x] No new TODOs introduced or explained herafter

v0d1ch committed:
Use production-grade webserver

ch1bo commented on CHANGELOG.md:
```diff
@@ -45,6 +45,9 @@ changes.
   + View the transactions in the server output encoded as CBOR
 
+- Replaced existing websocket server with production-grade one. This change also
+  enables serving HTTP API responses if needed.
```
Should: not mention http.. any webserver can do that and we are not actively adding this.

ch1bo commented on hydra-node/src/Hydra/Network.hs:
```diff
@@ -96,7 +97,7 @@ instance ToCBOR NodeId where
 data Host = Host
   { hostname :: Text
-  , port :: PortNumber
+  , port :: Port
```
Should: keep `PortNumber` as the data type we use everywhere

`PortNumber` is a dedicated and more suitable type than `type Port = Int` for describing port numbers. We should rather just convert to `Int` when we need it.

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -136,24 +140,31 @@ runAPIServer ::
+        `catch` onServerException con
+
+  httpApp _ respond =
+    respond $ responseLBS status400 [] "Currently Hydra supports only WebSocket requests."
```
Could: remove "currently" and either call this `hydra-node` or just not put a subject

```suggestion
    respond $ responseLBS status400 [] "only WebSocket connections supported"
```

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -136,24 +140,31 @@ runAPIServer ::
-
-      withPingThread con 30 (pure ()) $
-        race_ (receiveInputs con) (sendOutputs chan con txDisplay)
+    let serverSettings = setHost (ipToHostPreference host) $ setPort port $ defaultSettings
```
Should: use `IsString` instance of `HostPreference` instead of the additional `ipToHostPreference` code

```suggestion
    let serverSettings = defaultSettings & setHost (fromString $ show host) & setPort port
```

ch1bo reviewed (changes requested):
This change could have been more minimal. IMO the `PortNumber` type is better than the `Port` type alias. Also, the `Hydra.API.Host` module is not needed if you use `fromString`.

I would prefer both things changed before merging this. However I marked it as "Should", so if you see things different - let's discuss.

v0d1ch committed:
PR review comments

- Biggest change is: use PortNumber instead Port
- Remove Host module
- Adapt the changelog
- Change wording in http server

ch1bo commented on hydra-node/golden/RunOptions.json:
No need to regenerate this golden file now.

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -172,13 +173,17 @@ library
     , serialise
     , small-steps
     , stm
+    , streaming-commons
```
What are we using from this package?

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -172,13 +173,17 @@ library
     , typed-protocols-cborg           ==0.1.0.0
     , unliftio
+    , wai
+    , wai-websockets
```
Do we depend on `websockets` still if we use this package now?

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -35,13 +35,16 @@ import Network.WebSockets (
   sendTextData,
   sendTextDatas,
-  withPingThread,
+  withPingThread, defaultConnectionOptions, ConnectionException (..), sendClose, fromLazyByteString
```
Could: fix formatting

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -184,7 +194,21 @@ runAPIServer host port party tracer history callback responseChannel = do
+        throwIO pe
+      ue@(UnicodeException e) -> do
+        traceWith tracer (APIInvalidInput e "")
+        throwIO ue
```
Should: trace all exceptions

there can be any kind of exceptions raised by the server (not only https://hackage.haskell.org/package/websockets-0.12.7.3/docs/Network-WebSockets.html#t:ConnectionException). If we want to trace them, why not trace all of them?

If we just throw them, they will still show up on the stderr, but not as json.

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -184,7 +194,21 @@ runAPIServer host port party tracer history callback responseChannel = do
+    case exc of
+      -- ignoring 'ConnectionClosed' exceptions since they appear when client
+      -- doesn't send 'Close' meaning it doesn't close the connection cleanly
+      ConnectionClosed -> return ()
```
Should: log this case

ch1bo commented on hydra-node/src/Hydra/Options.hs:
```diff
@@ -168,11 +168,11 @@ instance Arbitrary RunOptions where
-    apiPort <- arbitrary
-    monitoringPort <- arbitrary
+    apiPort <- getPositive <$> arbitrary
+    monitoringPort <- oneof [Just . getPositive <$> arbitrary, pure Nothing]
```
Leftover `getPositive` AFAIS

ch1bo reviewed (approved):
Exception handling could be improved now, but it's not worse than before..  so approved.

v0d1ch committed:
Rebase master fix

v0d1ch committed:
PR review

v0d1ch committed:
PR review

ch1bo commented on hydra-node/src/Hydra/API/Server.hs:
```diff
@@ -136,24 +142,33 @@ runAPIServer ::
+    let serverSettings =
+          setHost (fromString $ show host) $
+            setPort (fromIntegral port) $
+              setOnException (\_ _ -> pure ()) defaultSettings
```
`APIConnectionError` and `APIHandshakeError` are unused now.

Should: Just trace any `SomeException` we get here (so it's not hidden).
```suggestion
              setOnException (\_ e -> traceWith tracer $ APIConnectionError { reason = show e}) defaultSettings
```
